hw2duhl.sml

```diff
@@ -0,0 +1,147 @@
+
+fun same_string(s1 : string, s2 : string) =
+    s1 = s2
+
+(* remove first instance of a string from a list *)
+fun all_except_option (str, strList) =
+    case strList of
+  [] => NONE
+      | x::xs => if
+        same_string(x, str) = true
+    then
+        SOME xs
+    else
+        case all_except_option(str, xs) of
+      NONE => NONE
+          | SOME(lst) => SOME(x::lst)
+
+fun get_substitutions1 (wordList, word) =
+    case wordList of
+  [] => []
+      | x::xs => let val nextSubs = get_substitutions1(xs, word)
+     in
+         case all_except_option(word, x) of
+       NONE => nextSubs
+           | SOME(lst) => lst @ nextSubs
+     end
+
+fun get_substitutions2 (words_lists, word) =
+    let fun f (words_lists, word, acc) =
+      case words_lists of
+    [] => acc
+        | x::xs => case all_except_option(word, x) of
+           NONE => f(xs, word, acc)
+         | SOME(lst) => f(xs, word, lst @ acc)
+    in
+  f(words_lists, word, [])
+    end
+
+fun similar_names(names_lists, fullName) =
+    case fullName of
+  {first, middle, last} => (case get_substitutions1(names_lists, first) of
+              [] => fullName::[]
+            | names => let fun f (firsts, middle, last) =
+                   case firsts of
+                 [] => []
+               | first::firsts' => {first=first, middle=middle, last=last}::f(firsts', middle, last)
+                 in
+               fullName::f(names, middle, last)
+                 end
+         )
+
+
+
+(* you may assume that Num is always used with values 2, 3, ..., 10
+   though it will not really come up *)
+datatype suit = Clubs | Diamonds | Hearts | Spades
+datatype rank = Jack | Queen | King | Ace | Num of int
+type card = suit * rank
```

Comment:
What's the difference between `type` and `datatype` again?

Reply:
type is an alias, datatype is a new type